Add constitutive relation for linear elasticity

nilsvu opened the pull request:
## Proposed changes

This PR adds first code for the linear elasticity system. Since parts of the XCTS equations are just a linear elasticity problem (a minimal distortion condition on the shift), this is a step towards implementing XCTS. We can also apply it to the thermal noise problem described, for instance, in https://arxiv.org/abs/1707.07774.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Rename Elasticity namespace and add tag for elastic stress

nilsvu committed:
Add test label for Elasticity

nilsvu committed:
Add constitutive relation for an isotropic elastic material

carmaza commented on src/Elliptic/Systems/Elasticity/Tags.hpp:
```diff
@@ -0,0 +1,60 @@
+ * field \f$\boldsymbol{u}\f$ in an elastic material that responds to an applied
+ * force \f$\boldsymbol{F}\f$. In this static approximation the equation of
+ * motion reduces to the elliptic equation \f[
+ * -\Nabla_i T^{ij} = F^j
```
`\nabla` here and in the definition of the strain below?

carmaza commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,85 @@
+ * \left(\delta^{ik}\delta^{jl} + \delta^{il}\delta^{jk}\right) \\
+ * \implies \quad T^{ij} = \lambda \mathrm{Tr}(S) \delta^{ij} + 2\mu S^{ij}
+ * \f]
+ * with the _Lamé parameter_ \f$\lambda\f$ and the _rigidity_ (or _shear
```
I think `Lam\'e` is preferred. There is another occurrence below.

nilsvu replied:
I think this was discussed but not resolved. I prefer the unicode.

geoffrey4444 replied:
I'm fine with unicode, but I seem to recall a consensus to avoid non-ascii in the code base

carmaza commented on src/Elliptic/Systems/Elasticity/Tags.hpp:
```diff
@@ -0,0 +1,60 @@
+ * \brief The amount of material deformation \f$\boldsymbol{u}(x)\f$ when a
+ * static force is applied.
+ */
+template <size_t Dim>
```
Should these tags be templated in `Frame` and possibly `DataType`, too? In that case, It might be useful to add a `TagsDeclarations` file as we do in `gr` and others.

nilsvu replied:
I can't say if that will become necessary any time soon. Therefore I'd prefer not to add that complication.

carmaza replied:
Sounds good.

geoffrey4444 replied:
Since these are elastostatic problems, I think an inertial frame should suffice wherever you need to specify a frame. Also, you'll always solve the equations on an extended domain, so I don't see the point of templating on DataType to support doubles.

carmaza commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,69 @@
+    const tnsr::I<DataType, 3, Frame>& /*x*/) const noexcept {
+  auto stress = make_with_value<tnsr::II<DataType, 3, Frame>>(strain, 0.);
+  for (size_t i = 0; i < 3; i++) {
+    for (size_t j = 0; j < 3; j++) {
```
Since `stress` is explicitly symmetric, and assuming that `strain` is correctly symmetrized, I believe you just need to loop over independent elements.

nilsvu replied:
I wasn't sure about that. @nilsdeppe can you confirm?

nilsdeppe replied:
Yep, I can confirm

carmaza commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,85 @@
+      "material."};
+
+  IsotropicHomogeneous() = default;
+  IsotropicHomogeneous(const IsotropicHomogeneous&) = default;
```
We've been `delete`ing this and the one below in other classes.

carmaza commented on tests/Unit/PointwiseFunctions/Elasticity/ConstitutiveRelations/Test_IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,27 @@
+
+SPECTRE_TEST_CASE("Unit.Elasticity.ConstitutiveRelations.IsotropicHomogeneous",
+                  "[PointwiseFunctions][Unit][Elasticity]") {
+  const Elasticity::ConstitutiveRelations::IsotropicHomogeneous<3> relation{1.,
```
Try less trivial values for the member variables, especially when checking the math below.

kidder commented:
@geoffrey4444 please review

nilsdeppe commented:
@geoffrey4444 please review this.

nilsvu commented:
@carmaza Addressed your comments, please check the fixups

carmaza commented on tests/Unit/PointwiseFunctions/Elasticity/ConstitutiveRelations/Test_IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,48 @@
+  const Elasticity::ConstitutiveRelations::IsotropicHomogeneous<3> relation{
+      incompressibility, rigidity};
+  test_serialization(relation);
+
```
Depending on how you'll use `ConstitutiveRelations`, you might also want to add a `test_move` and a `test_from_options` as we do in hydro analytic data.

nilsvu replied:
good point, added those tests

carmaza commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,95 @@
+
+ private:
+  double incompressibility_{};
+  double rigidity_{};
```
We've been setting these equal to `std::numeric_limits<double>::signaling_NaN()`. You'll need to `#include <limits>` before IWYU complains.

carmaza reviewed (commented):
Looks good. I left two extra (minor) comments, but that's all from me.

nilsvu commented:
Thanks @carmaza, I fixed those too. Could you check the updated `Test_IsotropicHomogeneous.cpp`? I squashed it, since the file is very small.

carmaza commented:
You're welcome, @nilsleiffischer ! Looks good to me.

kidder commented:
@geoffrey4444 🔔

geoffrey4444 commented on src/Elliptic/Systems/Elasticity/Tags.hpp:
```diff
@@ -0,0 +1,60 @@
+ * \f]
+ * where the stress \f$T^{ij}\f$ is related to the strain
+ * \f$S_{ij}=\nabla_{(i}u_{j)}\f$ by a linear constitutive relation
+ * \f$T^{ij}=\chi^{ijkl}S_{kl}\f$ that describes the elastic properties of the
```
I prefer \f$Y_{ijkl}\f$ for the Young's tensor

nilsvu replied:
sure, makes sense to be consistent with at least _one_ paper 😄 so let's make that yours

geoffrey4444 commented on src/Elliptic/Systems/Elasticity/Tags.hpp:
```diff
@@ -0,0 +1,60 @@
+ */
+template <size_t Dim>
+struct Displacement : db::SimpleTag {
+  using type = tnsr::I<DataVector, Dim>;
```
Most thermal-noise applications I know of use an orthonormal basis, so it's simplest to just keep all the indices on everything down. But maybe this doesn't fit well with how this will generalize to XCTS?

nilsvu replied:
Indeed I think the current setup of tensor types makes most sense for XCTS (and also elasticity). Not 100% sure, but my current understanding.

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * \brief An isotropic and homogeneous material
+ *
+ * \details For an isotropic and homogeneous material the linear constitutive
+ * relation \f$T^{ij}=\chi^{ijkl}S_{kl}\f$ reduces to
```
Again, I prefer \f$Y_{ijkl}\f$ for the Young's tensor here

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * \f[
+ * \chi^{ijkl} = \lambda \delta^{ij}\delta^{kl} + \mu
+ * \left(\delta^{ik}\delta^{jl} + \delta^{il}\delta^{jk}\right) \\
+ * \implies \quad T^{ij} = \lambda \mathrm{Tr}(S) \delta^{ij} + 2\mu S^{ij}
```
Can you please cit an equation for each relation here? Just below Eq. (11.19) in Thorne and Blandford, they have the same relation as you for T^{ij} except the right hand side differs from this by a sign, as does the RHS of Thorne and Blandford Eq. (11.18). Are you sure about the sign here? I would have thought, as Thorne and Blandford argue, that you need the minus signs so the material resists deformation when stressed, analogously to the minus sign in Hooke's law.

nilsvu replied:
I'll have to check the book, but I think the minus sign you mean you can find in the documentation of the `Elasticity` namespace (`Elasticity/Tags.hpp`):
$-\nabla_i T^{ij} = F^j$
I considered it part of the e.o.m. as opposed to the constitutive relation.

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * relation \f$T^{ij}=\chi^{ijkl}S_{kl}\f$ reduces to
+ * \f[
+ * \chi^{ijkl} = \lambda \delta^{ij}\delta^{kl} + \mu
+ * \left(\delta^{ik}\delta^{jl} + \delta^{il}\delta^{jk}\right) \\
```
I agree with this equation (it has all the right symmetries, and I use it in my thermal noise code), but I can't seem to find a good reference for it. Do you have one you can cite here?

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * use the _incompressibility_ (or _bulk modulus_) \f[K=\lambda +
+ * \frac{2}{3}\mu\f] instead of the Lamé parameter. In this parametrization the
+ * stress-strain relation \f[ T^{ij} = K \mathrm{Tr}(S) \delta^{ij} +
+ * 2\mu\left(S^{ij} - \frac{1}{3}\mathrm{Tr}(S)\delta^{ij}\right)
```
Looks OK except RHS has the wrong sign. Also, might be worth defining the symmetric-trace-free part of the strain, which is the shear tensor, as in Thorne and Blandford Eq. (11.18). This makes the equation much simpler.

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * \f]
+ * decomposes into a scalar and a traceless part. Parameters also often used in
+ * this context are the _Young's modulus_
+ * \f[E=\frac{9K\mu}{3K+\mu}=\frac{\mu(3\lambda+2\mu)}{\lambda+\mu}\f] and the
```
First part of this equality is correct. Again, an example where this is in the literature would be nice

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * this context are the _Young's modulus_
+ * \f[E=\frac{9K\mu}{3K+\mu}=\frac{\mu(3\lambda+2\mu)}{\lambda+\mu}\f] and the
+ * _Poisson ratio_
+ * \f[\nu=\frac{3K-2\mu}{6K+2\mu}=\frac{\lambda}{2(\lambda+\mu)}\f].
```
First equality is correct. I didn't check the algebra on the others

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * \f[\nu=\frac{3K-2\mu}{6K+2\mu}=\frac{\lambda}{2(\lambda+\mu)}\f].
+ * Inversely, these relations read:
+ * \f{align}
+ * K &=\frac{E}{3-6\nu} \\
```
Equation is correct

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+ * \f{align}
+ * K &=\frac{E}{3-6\nu} \\
+ * \lambda &=\frac{E\nu}{(1+\nu)(1-2\nu)} \\
+ * \mu &=\frac{E}{2(1+\nu)}
```
Equation is correct

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+template <size_t Dim>
+class IsotropicHomogeneous {
+ public:
+  struct Incompressibility {
```
Why not call this BulkModulus? That would be much clearer what it is, I think.

nilsvu replied:
agreed, changed

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+    static constexpr OptionString help = {"The bulk modulus of the material"};
+  };
+
+  struct Rigidity {
```
Why not call this ShearModulus?

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -0,0 +1,104 @@
+  template <typename DataType, typename Frame>
+  tnsr::II<DataType, Dim, Frame> stress(
+      const tnsr::ii<DataType, Dim, Frame>& strain,
+      const tnsr::I<DataType, Dim, Frame>& x) const noexcept;
```
Isn't it true that all you need is the strain and the elastic moduli to get the stress tensor everywhere? Why compute it at a single point?

nilsvu replied:
We may want to implement inhomogeneous constitutive relations, which would need this

geoffrey4444 commented on tests/Unit/PointwiseFunctions/Elasticity/ConstitutiveRelations/Test_IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,68 @@
+
+namespace {
+
+void test_implementation(const double incompressibility,
```
I suggest testing this analytically, say for a known isotropic material

nilsvu replied:
added some tests, please check the fixup

geoffrey4444 commented on tests/Unit/PointwiseFunctions/Elasticity/ConstitutiveRelations/Test_IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,68 @@
+  test_implementation(56.6667, 5.86207);
+  // Wood (transverse direction): E=1, nu=0.79
+  test_implementation(-0.574713, 0.27933);
+  // Parameters used for solving a Bowen-Yorck momentum constraint:
```
York

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,74 @@
+template <typename DataType, typename Frame>
+tnsr::II<DataType, 3, Frame> IsotropicHomogeneous<3>::stress(
+    const tnsr::ii<DataType, 3, Frame>& strain,
+    const tnsr::I<DataType, 3, Frame>& /*x*/) const noexcept {
```
I'm confused why you are passing x at all...does this have to do with the elliptic solver somehow?

nilsvu replied:
see [above](https://github.com/sxs-collaboration/spectre/pull/1298#discussion_r258469238)

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,74 @@
+  }
+  auto trace_term = make_with_value<DataType>(strain, 0.);
+  for (size_t i = 0; i < 3; i++) {
+    trace_term += strain.get(i, i);
```
You can probably use an existing trace() function here

nilsvu replied:
I think those require passing in a metric

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,74 @@
+    const tnsr::I<DataType, 3, Frame>& /*x*/) const noexcept {
+  auto result = make_with_value<tnsr::II<DataType, 3, Frame>>(strain, 0.);
+  for (size_t i = 0; i < 3; i++) {
+    for (size_t j = 0; j <= i; j++) {
```
We normally don't account for symmetry in spectre loops, do we?

nilsvu replied:
please see [above](https://github.com/sxs-collaboration/spectre/pull/1298#discussion_r252857962)

geoffrey4444 commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.cpp:
```diff
@@ -0,0 +1,74 @@
+  for (size_t i = 0; i < 3; i++) {
+    result.get(i, i) += trace_term;
+  }
+  return std::move(result);
```
To help me understand, what is the std::move doing here?

nilsdeppe replied:
Nothing good, this is always an incorrect usage of `std::move`. It prevents copy elision (which will be guaranteed in C++17 and all compilers already do it). (I'm surprised clang-tidy didn't complain about this)

nilsvu committed:
Add tag for the constitutive relation of an elastic material

nilsvu commented:
@geoffrey4444 @carmaza I have added a bunch of fixups to address comments, and also to add a `ConstitutiveRelation` base class. This is similar to how equations of state are implemented. Please look at the fixups and let me know when to squash them :)

carmaza commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp:
```diff
@@ -75,30 +99,51 @@ class IsotropicHomogeneous {
+      noexcept override;
+
+  /// The bulk modulus (or incompressibility) \f$K\f$
+  double bulk_modulus() const noexcept;
```
I believe the consensus for analytic solutions was to remove these getters and have `operator==` be `friend`.

nilsvu replied:
I need these to compute analytic solutions such as #1374

carmaza reviewed (commented):
Looks good. Should be ready for a core developer to look at.

nilsvu commented:
@geoffrey4444 I added references and, to be consistent with Thorne/Blandford, switched the sign of the constitutive relation. Please check the fixup commits and let me know when to squash. Thanks!

nilsdeppe commented:
@geoffrey4444 please look at this.

nilsvu commented:
Thanks for the feedback @geoffrey4444 and @carmaza! I squashed and rebased, so this should be ready for a final review @sxs-collaboration/spectre-core-devs

nilsvu commented:
ping @sxs-collaboration/spectre-core-devs

nilsvu commented:
Also @geoffrey4444 and @carmaza could you give your approval so that this can be merged?

nilsvu commented:
ping @sxs-collaboration/spectre-core-devs @geoffrey4444

kidder commented:
@geoffrey4444 are you happy with this?

kidder commented:
@geoffrey4444 promised to look at this today
